### PR TITLE
Setup FlatFieldStep to handle user-provided flat field instead of calculating from references

### PR DESCRIPTION
**Pull request by stscieisenhamer**
Review: A rebase with master was necessary to get some previous basic changes to `calwebb_spec2` incorporated, so there is a lot of noise here. The relevant files for this feature are:

calwabb_spec2.py
flat_field_step.py
flat_field.py
regtest/test_nirspec* <-- that involve spec2 and flatfield

Basically this is allow a user-specified flat field, which will most likely be a flat field save from `save_interpolated_flats`, to be used. For most modes, this would be equivalent to replacing the `flat` reference file. However, for NIRSpec, providing such a flat skips the re-calculation to create the interpolated flats.

### CODE REVIEW

**Comment by hbushouse** [comment]
Oh, it's missing a change log entry.

File: jwst/flatfield/flat_field.py [hunk 1]
```diff
@@ -302,126 +318,39 @@ def nirspec_fs_msa(output_model, f_flat_model, s_flat_model, d_flat_model,
...
-                dummy_flat.xstart = slit.xstart
-                dummy_flat.xsize = slit.xsize
-                dummy_flat.ystart = slit.ystart
-                dummy_flat.ysize = slit.ysize
-                dummy_flat.wavelength = np.zeros_like(slit.data)
-                flat_slits.append(dummy_flat)
-
-                continue
+        if user_supplied_flat is not None:
+            slit_flat = user_supplied_flat.slits[slit_idx]
```
[the remark this answers is not in the document]
Comment (stscieisenhamer) [reply jwst/flatfield/flat_field.py:329]: Yes, at this point the form of the flats really needs to match. I'm not sure there is any other way of handling it, besides sort of the master background operation anyways where one converts a generic flat to the desired flat. Everything that constructs the flat for the 3-tuple is going to be data-specific. If this method is continued to be implemented, but user warnings will need be documented.

File: jwst/flatfield/flat_field.py [hunk 2]
```diff
@@ -1655,3 +1429,383 @@ def interpolate_flat(image_flat, image_dq, image_err, image_wl, wl):
     flat_2d[:, :] = np.where(flat_dq > 0, 1., flat_2d)
 
     return flat_2d.astype(image_flat.dtype), flat_dq, flat_err
+
+
+def flat_for_nirspec_ifu(output_model, f_flat_model, s_flat_model, d_flat_model,
```
[the remark this answers is not in the document]
Comment (stscieisenhamer) [reply jwst/flatfield/flat_field.py:1434]: Local regressions seem fine. Did actually run across some minor bugs, mostly with error arrays not getting passed into the interpolated flats. Running an official regression to see what is up.

**Comment by stscieisenhamer** [comment]
Removing this request in favor of a more comprehensive one.

**Pull request closed without merging**